r1cs/gadgets/std/src/groups/group_vec.rs

```
@@ -0,0 +1,402 @@
+    fn to_bits<CS: ConstraintSystemAbstract<ConstraintF>>(&self, mut cs: CS) -> Result<Vec<Boolean>, SynthesisError> {
+        let mut res_vec = Vec::new();
+        for (i, el) in self.iter().enumerate() {
+            let el_to_bits = el.to_bits(cs.ns(|| format!("convert element {} to bits", i)))?;
+            res_vec.extend_from_slice(el_to_bits.as_slice());
+        }
+        Ok(res_vec)
+    }
+
+    fn to_bits_strict<CS: ConstraintSystemAbstract<ConstraintF>>(&self, mut cs: CS) -> Result<Vec<Boolean>, SynthesisError> {
+        let mut res_vec = Vec::new();
+        for (i, el) in self.iter().enumerate() {
+            let el_to_bits = el.to_bits_strict(cs.ns(|| format!("convert element {} to bits with check", i)))?;
+            res_vec.extend_from_slice(el_to_bits.as_slice());
+        }
+        Ok(res_vec)
+    }
+}
+
+impl<ConstraintF: Field, G: Group, GG: GroupGadget<G, ConstraintF>> CondSelectGadget<ConstraintF> for GroupGadgetVec<ConstraintF, G, GG> {
+    fn conditionally_select<CS: ConstraintSystemAbstract<ConstraintF>>(mut cs: CS, cond: &Boolean, first: &Self, second: &Self) -> Result<Self, SynthesisError> {
+        let mut res_vec = Vec::new();
+        for (i,(el1, el2)) in first.iter().zip(second.iter()).enumerate() {
+            res_vec.push(GG::conditionally_select(cs.ns(|| format!("cond select element {}", i)), cond, el1, el2)?);
+        }
+        let zero = GG::zero(cs.ns(|| "alloc zero"))?;
+        if first.len() > second.len() {
+            for (i, el) in first.iter().enumerate().skip(second.len()) {
+                res_vec.push(GG::conditionally_select(cs.ns(|| format!("cond select element {}", i)), cond, el, &zero)?)
+            }
+        } else {
+            for (i, el) in second.iter().enumerate().skip(first.len()) {
+                res_vec.push(GG::conditionally_select(cs.ns(|| format!("cond select element {}", i)), cond, &zero, &el)?)
+            }
+        };
```

Mmm honestly, considering the way we use it, it might be dangerous to implicitly perform the conditional selection even if the two have different length. Please check if this flexibility is really useful and needed. If not, I would throw an error if they are not of the same length and keep only lines 113-116. If you think they are, I think the best is to pad one of them explicitly in the caller code, but we can discuss it.

---

In general, I would keep this gadget as specific as possible, to limit weird generic usage, with the promise to deprecate it once (and if) we will refactor PC taking into account that commitments are always segmented

---

Since I change the zero representation to be the empty vector, now it is possible to let the conditionally select return an error if the two input vectors have different length, except for the case when at least one of them is the empty vector, which needs to be explicitly handled. Therefore, I removed this portion of the code and I added another piece which deals with either of the two vectors being the empty one before checking that the two vectors have the same length